### PR TITLE
Add combine_with_exported_module and a type cast fix for constants.

### PR DESCRIPTION
**Pull request by schungx**

### CODE REVIEW

**Comment by schungx** [comment]
@jhwgh1968   I'm beginning to study the proc macros code.  This is a small attempt to add `combine_with_exported_module` which flatten-generate module functions into an existing module.

I also fix a small problem with types when processing constants in modules.  The expression was not originally guaranteed to be of the same type as the constant definition.  I added a type cast to make sure that it is the same.

**Comment by schungx** [comment]
I can see that I need to fix the tests in `src/test/module.rs` to match the actually generated code.

I'll do that gradually.

**Comment by jhwgh1968** [comment]
Looking at the PR, I would say the new function does no harm... except there are a couple changes to the unit tests (under `src/test`) that surprise me. I would like to understand those.

Since more test changes are coming, I will wait until the PR passes CI before doing a detailed review. That context will help.

**Comment by schungx** [comment]
> Looking at the PR, I would say the new function does no harm... except there are a couple changes to the unit tests (under `src/test`) that surprise me. I would like to understand those.

I think the main one is that I added an extra parameter to constants, keeping its type.

```rust
mod some_mod {
    const foo: INT = 42;        // 42 is volatile
    const foo2: INT = 42 as INT;     // 42 is always INT
}
```

In the case above, `42` will be cast to the system default integer type, not `INT` as specified.  That's because, internally, the macro creates a `set_var` call with just the expression in it, throwing away the type.

I changed it to cast it to the right type:

```rust
m.set_var("foo", (42) as INT);
```

**Comment by schungx** [comment]
@jhwgh1968   Yay!  Tests passed!

Basically, I added `rhai_generate_into_module` which registers functions into an existing module.  Then `rhai_generate` is modified to simply call that.  Unfortunately that means `src/test/module.rs` needs to be extensively modified to match the new generated tokens.

**Comment by jhwgh1968** [comment]
Thanks to the extra permissions you've given me, I can simply clean up your branch, rather than trying to explain another idea I had! :sparkles:

I will review that version once I push it.

File: codegen/src/rhai_module.rs
Comment (jhwgh1968) [inline codegen/src/rhai_module.rs:9]: Why is this a `Box` instead of just a `syn::Type`? If it just for the `clone`, then see my code suggestion earlier.

File: codegen/src/module.rs
Comment (jhwgh1968) [inline codegen/src/module.rs:143]:
```suggestion
                        ref ty,
```

File: codegen/src/test/function.rs
Comment (jhwgh1968) [inline codegen/src/test/function.rs:240]: I think this is more difficult to read than the original, because I have to think about the returns and the full conditional. Was this a suggestion by clippy?

**Comment by schungx** [comment]
If possible, maybe clean up the commit history?  I screwed up now there are many duplicated commits...  Don't know how to fix it...

File: codegen/src/test/function.rs
Comment (jhwgh1968) [inline codegen/src/test/function.rs:258]: Believe it or not, I would prefer that you delete the entire logic building `actual_diff` and `expected_diff` if this is to get rid of an unused code warning.

I put this in originally, but haven't used it in a long time. It is more visually difficult for me to look at than I thought. It is easier for me to see the character count, and then read the mostly aligned `assert_eq` output, counting the characters from the start.

If you'd like to see what I mean, make a test fail on purpose.

**Review by jhwgh1968 — CHANGES_REQUESTED** [review]
I am still thinking there is a better way to handle constants than duplicating them all, but first, some other comments.

File: codegen/src/lib.rs
Comment (schungx) [reply codegen/src/lib.rs:157]: Laziness...  :-D   I originally put it in just to shut the warnings up.  Then I realized I need to change all the tests.  Then I changed all the tests, but then figured out that I can simply make the variable `_`...

And then laziness took over...

File: codegen/src/rhai_module.rs
Comment (schungx) [reply codegen/src/rhai_module.rs:9]: No particular reason, except that I copied it from some other struct...  Not very familiar with `syn::Type` to know the difference.

Comment (schungx) [reply codegen/src/rhai_module.rs:9]: I'll fix it.

File: codegen/src/module.rs
Comment (schungx) [reply codegen/src/module.rs:143]: I'll fix `ExportedConst` not to take a `Box`.

File: codegen/src/test/function.rs
Comment (schungx) [reply codegen/src/test/function.rs:240]: No.  `counter` has the wrong value.  It is 50 characters too many when I turned on the 50-character diff print-out.

That's because the `inspect` clause is run even for the diff print-out.  `counter += 1` must be moved inside `skip_while`.

File: codegen/src/test/function.rs
Comment (schungx) [reply codegen/src/test/function.rs:258]: No problem.  I can do that.

It was just a matter of preference.  I found an entire screenful of diffs, especially for vary large ones like the deeply-nested test, hard to find the error position, while the 50-character display shows it clearly and concisely.

I originally turned it back on to let me figure out where I went wrong with the last, deep-nesting test.  I suppose it is not very necessary for shorter token streams.

I'll put this under a comment block so that we can simply turn it back on later if needed.

File: codegen/src/lib.rs
Comment (schungx) [reply codegen/src/lib.rs:157]: Laziness...  :-D

Comment (schungx) [reply codegen/src/lib.rs:157]: Will fix.

**Comment by schungx** [comment]
> I am still thinking there is a better way to handle constants than duplicating them all, but first, some other comments.

Maybe keep the constant definition, but then use it in the `set_var` call instead of duplicating the code with a cast?

**Comment by schungx** [comment]
> I am still thinking there is a better way to handle constants than duplicating them all, but first, some other comments.

Changed `set_var(#const_name, (#const_expr) as #const_type)` to `set_var(#const_name, #const_value)` where `const_value` is `Ident(const_name)`.

**Comment by jhwgh1968** [comment]
I have reviewed again, and cleaned up the extra merge commit. LGTM. Once CI passes, I will approve.

**Comment by jhwgh1968** [comment]
As a side note, I just found this: [Git rebase in Visual Studio Code?](https://stackoverflow.com/questions/51381826/git-rebase-in-visual-studio-code)

They say you can avoid the merge commits with a 3rd party extension [which appears to be free](https://marketplace.visualstudio.com/items?itemName=eamodio.gitlens).

**Review by jhwgh1968 — APPROVED** [review]

**Comment by schungx** [comment]
Yes, I have been using `git pull --rebase` as you've taught.  However, it seems like the commit history lists keeps going longer on this one...  

It must have been Visual Studio Code that is doing this, but I"m not sure.  It could also be something that I've done wrong...

**Comment by jhwgh1968** [comment]
Whoa, hang on, you just added another merge commit! I was actually hoping you would wait to merge it before you did anything else.

Should I fix it again? Or are you trying to add something else?

**Comment by schungx** [comment]
Yup.  I just pressed "Sync" and it added a merge commit, for some weird reason.  :-(

Please give this a fix.  I'll actually delete my local branch and pull from here.

**Comment by jhwgh1968** [comment]
Alright, it's fixed!

**Comment by schungx** [comment]
Thanks!  I deleted my local branch and then `git checkout -b plugins_dev origin/plugins_dev` and it seems to be working now...

**Comment by schungx** [comment]
Hhhmmm... for some reason, three `no-std` tests remain pending for a long time.

I'm just merging this...

**Pull request merged**
